fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1346,6 +1352,242 @@ def assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None
                 traceback.print_stack()
             raise ValueError(msg)
 
+    # Optim State dict functions
+
+    def consolidate_optim_state_dict(self, optim: torch.optim.Optimizer, recipient_rank: int = 0) -> None:
+        """Update the consolidated state_dict list, one per rank.
+
+        Arguments:
+            recipient_rank (int): on which rank to materialize the full state dict.
+            -1 is a special value, which means that all ranks should have the state
+
+        .. warning: This needs to be called on all replicas"""
+        _default_device = torch.device("cuda")
+        # NOTE(SS): we do not support param groups yet, as they seem to break FSDP
+
+        # Pull the sharded state from all the other replicas
+        # Store all the states in order, rank by rank
+        should_collect_state = self.rank == recipient_rank or recipient_rank == -1
+        should_send_state = (self.rank != recipient_rank and recipient_rank != -1) or recipient_rank == -1
+        _all_optimizer_states: List[Dict[str, Any]] = []
+        for rank in range(self.world_size):
+            if rank == self.rank:
+                sd = optim.state_dict()
+                sd["num_padded"] = [m.num_padded for m in self.modules() if isinstance(m, FullyShardedDataParallel)]
+                if should_collect_state:
+                    _all_optimizer_states.append(
+                        recursive_copy_to_device(sd, non_blocking=True, device=torch.device("cpu"))
```

Comment:
remove `non_blocking`, it will actually be blocking no matter what, since GPU->CPU transfer can only be non-blocking if the destination is pinned memory.

Roughly speaking (not tested for syntax):
```python
x = torch.ones(5).cuda()
buffer = torch.zeros(5)
buffer.copy_(x, non_blocking=True)  # blocking
buffer.pin_memory_()
buffer.copy_(x, non_blocking=True)  # non-blocking
```

Reply:
Thx for teaching!